Allow specifying denoising_start and denoising_end as integers representing the discrete timesteps, fixing the XL ensemble not working for many schedulers

AmericanPresidentJimmyCarter opened the pull request:
# What does this PR do?

Partially addresses #4060 and fixes XL ensemble-of-experts for most schedulers.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Example

```py
from diffusers import StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline
from diffusers.schedulers import DPMSolverMultistepScheduler
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

hf_cache_dir = '/path/to/cache'

pipe = StableDiffusionXLPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-base-0.9", variant="fp16", use_safetensors=True,
    torch_dtype=torch.bfloat16,
)
pipe.enable_model_cpu_offload(1)
pipe.enable_vae_slicing()
pipe.enable_vae_tiling()


refiner = StableDiffusionXLImg2ImgPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-refiner-0.9", use_safetensors=True, variant="fp16",
    torch_dtype=torch.bfloat16,
)
refiner.enable_model_cpu_offload(1)
refiner.enable_vae_slicing()
refiner.enable_vae_tiling()

with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
    use_refiner = True

    random_generator = torch.Generator()
    random_generator.manual_seed(12345)
    prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
    image = pipe(prompt=prompt, output_type="pil",
        num_inference_steps=30, generator=random_generator).images[0]
    image.save('spaceman1.png')

    random_generator.manual_seed(12345)
    image = pipe(prompt=prompt, output_type="latent" if use_refiner else "pil",
        denoising_end=200, num_inference_steps=30, generator=random_generator).images[0]
    image = refiner(prompt=prompt, image=image[None, :], denoising_start=200, num_inference_steps=30).images[0]
    image.save('spaceman2.png')

pipe.register_modules(scheduler=scheduler)
refiner.register_modules(scheduler=scheduler)

with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
    use_refiner = True

    random_generator = torch.Generator()
    random_generator.manual_seed(12345)
    prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
    image = pipe(prompt=prompt, output_type="pil",
        num_inference_steps=100, generator=random_generator).images[0]
    image.save('spaceman1_dpms.png')

    random_generator.manual_seed(12345)
    image = pipe(prompt=prompt, output_type="latent" if use_refiner else "pil",
        denoising_end=200, num_inference_steps=100, generator=random_generator).images[0]
    image = refiner(prompt=prompt, image=image[None, :], denoising_start=200, num_inference_steps=100).images[0]
    image.save('spaceman2_dpms.png')
```

Original:
![spaceman100_1_dpms](https://github.com/huggingface/diffusers/assets/110263573/ec9f53b2-157b-4186-ae67-7bc746ad1d27)

With refinement model:
![spaceman100_2_dpms](https://github.com/huggingface/diffusers/assets/110263573/4a4694c9-e1ec-4551-a3a7-fbf61b00d2d5)


## Who can review?

@pcuenca @patrickvonplaten @sayakpaul

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4115). All of your documentation changes will be reflected on that endpoint.

AmericanPresidentJimmyCarter commented:
Here are the two images that need to be added to HF image repository, I'm not sure exactly how to do that:

lion_base_timesteps.png

![lion_base_timesteps](https://github.com/huggingface/diffusers/assets/110263573/aa60db72-fd75-4e6b-831c-7a91dcbb9be9)

lion_refined_timesteps.png

![lion_refined_timesteps](https://github.com/huggingface/diffusers/assets/110263573/919d872b-9c9f-46f8-b16e-b53d8b1df137)

AmericanPresidentJimmyCarter commented:
@n00mkrad This branch should be working.

sayakpaul commented:
@AmericanPresidentJimmyCarter please open a PR to https://huggingface.co/hf-internal-testing w.r.t https://github.com/huggingface/diffusers/pull/4115#issuecomment-1636846164. Let us know if you face any issues.

n00mkrad commented:
> @n00mkrad This branch should be working.

Sadly DPM++2M is still broken. The other samplers seem to work though.

Left: Euler Karras - Right: DPM++2M Karras
![image](https://github.com/huggingface/diffusers/assets/61149547/7e23b1f2-2d21-4bdf-8209-663b3cc2aeb8)

AmericanPresidentJimmyCarter commented:
> > @n00mkrad This branch should be working.
> 
> Sadly DPM++2M is still broken. The other samplers seem to work though.
> 
> 25 Steps, 70/30 base/refiner split.

The astronaut images above are with DPM++2M. Are you using the integer value `200` for the end and start timesteps as detailed above in the script? The float value of `0.30` works exactly the same as before.

n00mkrad commented:
> > > @n00mkrad This branch should be working.
> > 
> > 
> > Sadly DPM++2M is still broken. The other samplers seem to work though.
> > 25 Steps, 70/30 base/refiner split.
> 
> The astronaut images above are with DPM++2M. Are you using the integer value `200` for the end and start timesteps as detailed above in the script? The float value of `0.30` works exactly the same as before.

Yes.

Another issue is that the steps apparently get miscalculated - Denoise end 200 with 50 steps gets me 32 base + 10 refiner (= 42) steps, for example.

I'll check your script if there are any differences to mine.

n00mkrad commented:
I think you just didn't notice the quality difference because you ran 100 steps which is kinda overkill.

Try 25 steps with `300` timesteps.

AmericanPresidentJimmyCarter commented:
> EDIT: The wrong step count only appears to happen with non-Multistep schedulers.

Yeah, I noticed this on the main branch too. It's an unrelated bug and I will open an issue for it.

AmericanPresidentJimmyCarter commented:
> Positive: `photo of an astronaut standing in a jungle, muted colors, detailed, 8k, realistic, helmet reflections`
> Negative: `blurry, ugly, low quality, jpeg compression, drawing, sketch, painting, cartoon, face`
> Seed: `42`
> CFG Scale: `6.5`
> Resolution: `1024x1024`
> denoising_end/denoising_start: `300`

```py
from diffusers import StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline
from diffusers.schedulers import DPMSolverMultistepScheduler
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

hf_cache_dir = '/path/to/cache'

pipe = StableDiffusionXLPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-base-0.9", variant="fp16", use_safetensors=True,
    torch_dtype=torch.bfloat16,
)
pipe.enable_model_cpu_offload(0)
pipe.enable_vae_slicing()
pipe.enable_vae_tiling()


refiner = StableDiffusionXLImg2ImgPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-refiner-0.9", use_safetensors=True, variant="fp16",
    torch_dtype=torch.bfloat16,
)
refiner.enable_model_cpu_offload(0)
refiner.enable_vae_slicing()
refiner.enable_vae_tiling()

pipe.register_modules(scheduler=scheduler)
refiner.register_modules(scheduler=scheduler)

with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
    use_refiner = True

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    prompt = "photo of an astronaut standing in a jungle, muted colors, detailed, 8k, realistic, helmet reflections"
    n_prompt = "blurry, ugly, low quality, jpeg compression, drawing, sketch, painting, cartoon, face"
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="pil",
        num_inference_steps=25, generator=random_generator).images[0]
    image.save('spaceman25_1.png')

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="latent" if use_refiner else "pil",
        denoising_end=300, num_inference_steps=25, generator=random_generator).images[0]
    image = refiner(prompt=prompt, image=image[None, :], denoising_start=300, num_inference_steps=25).images[0]
    image.save('spaceman25_2.png')
```

![spaceman25_1](https://github.com/huggingface/diffusers/assets/110263573/51b9e888-531e-4280-bcbb-346900325876)

![spaceman25_2](https://github.com/huggingface/diffusers/assets/110263573/0f629a7c-a937-4453-ab55-05fdaa153fa1)

AmericanPresidentJimmyCarter commented:
> @AmericanPresidentJimmyCarter please open a PR to https://huggingface.co/hf-internal-testing w.r.t [#4115 (comment)](https://github.com/huggingface/diffusers/pull/4115#issuecomment-1636846164). Let us know if you face any issues.

The instructions on huggingface don't work for git after opening a pull request under "advanced mode" (using GIT) through your web UI. In particular, it gives the instructions to:

`git clone https://huggingface.co/datasets/huggingface/documentation-images`

However, when attempting to auth via https later to push my changes (`git push origin pr/137:refs/pr/137
`), I am prompted for a user/pass and auth, but that is always refused.

I was able to get around this by:

```bash
git remote add origin2 git@hf.co:datasets/huggingface/documentation-images
```

Then logging in with my HF token and `git push origin2 pr/137:refs/pr/137`. I am not sure who controls these docs at your organization but you might want to ask them to update this since it took me a while to figure out.

Here is the PR, I hope I did everything else correctly:

https://huggingface.co/datasets/huggingface/documentation-images/discussions/137

sayakpaul commented:
Cc: @osanseviero @Wauplin for the bit on instructions mentioned by @AmericanPresidentJimmyCarter. 

Let's get this PR reviewed first and we can then merge https://huggingface.co/datasets/huggingface/documentation-images/discussions/137/. Hope that works.

n00mkrad commented:
@AmericanPresidentJimmyCarter 
My bad, I only set the scheduler for the base pipeline, but not for the refiner. When setting the scheduler on both pipelines, it works as intended.

Wauplin commented:
(About the PR git instructions, I opened an [internal issue](https://github.com/huggingface/moon-landing/pull/7031) -private link- to make it clearer cc @osanseviero.

FYI @AmericanPresidentJimmyCarter the easiest way to use git is to log in with `huggingface-cli login` in your terminal. This will set git credentials for your machine. It is documented [here](https://huggingface.co/docs/hub/repositories-getting-started#requirements) but a reminder in the PR steps wouldn't hurt #4001 )

patrickvonplaten commented:
To better understand what's going on here.

To my eyes, there is no issue with the current implementation of `denoising_start` and `denoising_end`. The following works just fine:

```py
from diffusers import DPMSolverMultistepScheduler
from diffusers import DiffusionPipeline

import torch

random_generator = torch.Generator()
random_generator.manual_seed(12345)

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

model_id = "stabilityai/stable-diffusion-xl-base-0.9"
pipe_high_noise = DiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16, variant="fp16", use_safetensors=True, local_files_only=True)

pipe_high_noise.scheduler = scheduler
pipe_high_noise.to("cuda")

pipe_low_noise = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-0.9", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")

pipe_low_noise.scheduler = scheduler
pipe_low_noise.to("cuda")

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"

num_inference_steps = 40
high_noise_frac = 0.8

image = pipe_high_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_end=high_noise_frac, output_type="latent").images
image = pipe_low_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_start=high_noise_frac, image=image).images[0]
```

So I think instead there is a problem with how we initialize the scheduler. Some schedulers are not state-less (like DPM) and here we probably need the exact same scheduler twice.

AmericanPresidentJimmyCarter commented:
To properly use the XL Refiner in an ensemble of experts, you need to use the base unet in timesteps 200-1000 and the refiner on timesteps 0-200, since the refiner is finetuned on steps 0-200 exclusively.

The current implementation does not work correctly because instead of targeting timesteps, it targets "strength", which is an arbitrary proportion of timesteps based on a fraction which may nor may not be the set of timesteps described above.

patrickvonplaten commented:
I'm getting pretty convincing results with it:
![aaa_1 (1)](https://github.com/huggingface/diffusers/assets/23423619/df76f6d4-9e74-4543-9b68-3c014566ab97)

patrickvonplaten commented:
@AmericanPresidentJimmyCarter, could you try whether the following works for you:

```py
#!/usr/bin/env python3
from diffusers import DPMSolverMultistepScheduler, DiffusionPipeline, EulerDiscreteScheduler, StableDiffusionPipeline, KDPM2DiscreteScheduler, StableDiffusionImg2ImgPipeline, HeunDiscreteScheduler, KDPM2AncestralDiscreteScheduler, DDIMScheduler, HeunDiscreteScheduler, DEISMultistepScheduler
from diffusers import DiffusionPipeline
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

model_id = "stabilityai/stable-diffusion-xl-base-0.9"
pipe_high_noise = DiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16, variant="fp16", use_safetensors=True, local_files_only=True)
pipe_high_noise.scheduler = scheduler
pipe_high_noise.to("cuda")

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

pipe_low_noise = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-0.9", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
pipe_low_noise.scheduler = scheduler
pipe_low_noise.to("cuda")

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"


random_generator = torch.Generator()
random_generator.manual_seed(0)

num_inference_steps = 100
high_noise_frac = 0.8

image = pipe_high_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_end=high_noise_frac, output_type="latent").images
image = pipe_low_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_start=high_noise_frac, image=image).images[0]
```
?

AmericanPresidentJimmyCarter commented:
> As far as I can see the following:
> 
> ```python
>     image = pipe(prompt=prompt, output_type="latent" if use_refiner else "pil",
>         denoising_end=200, num_inference_steps=100, generator=random_generator).images[0]
>     image = refiner(prompt=prompt, image=image[None, :], denoising_start=200, num_inference_steps=100).images[0]
> ```
> 
> is exactly the same as currently doing:
> 
> ```python
> num_inference_steps = 100
> high_noise_frac = 0.8
> 
> image = pipe_high_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_end=high_noise_frac, output_type="latent").images
> image = pipe_low_noise(prompt=prompt, num_inference_steps=num_inference_steps, denoising_start=high_noise_frac, image=image).images[0]
> ```

No, it is absolutely not the same thing. You can observe this by simply adding the following line to print the timesteps that are selected in the pipeline.

```diff
        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            print('timesteps', timesteps)
```

For both the StableDiffusionXLPipeline and StableDiffusionXLImg2ImgPipeline.

Here is the code for you to verify:

```py
from diffusers import StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline
from diffusers.schedulers import DPMSolverMultistepScheduler
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)

hf_cache_dir = '/path/to/cache'

pipe = StableDiffusionXLPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-base-0.9", variant="fp16", use_safetensors=True,
    torch_dtype=torch.bfloat16,
)
pipe.enable_model_cpu_offload(0)
pipe.enable_vae_slicing()
pipe.enable_vae_tiling()


refiner = StableDiffusionXLImg2ImgPipeline.from_pretrained(
    f"{hf_cache_dir}/stable-diffusion-xl-refiner-0.9", use_safetensors=True, variant="fp16",
    torch_dtype=torch.bfloat16,
)
refiner.enable_model_cpu_offload(0)
refiner.enable_vae_slicing()
refiner.enable_vae_tiling()

pipe.register_modules(scheduler=scheduler)
refiner.register_modules(scheduler=scheduler)

with torch.no_grad(), torch.amp.autocast('cuda', dtype=torch.bfloat16):
    use_refiner = True

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    prompt = "photo of an astronaut standing in a jungle, muted colors, detailed, 8k, realistic, helmet reflections"
    n_prompt = "blurry, ugly, low quality, jpeg compression, drawing, sketch, painting, cartoon, face"
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="pil",
        num_inference_steps=100, generator=random_generator, guidance_scale=6.5).images[0]
    image.save('spaceman100_1.png')

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="latent" if use_refiner else "pil",
        denoising_end=200, num_inference_steps=100, generator=random_generator, guidance_scale=6.5).images[0]
    image = refiner(prompt=prompt, image=image[None, :], denoising_start=200, num_inference_steps=100,
        guidance_scale=6.5).images[0]
    image.save('spaceman100_2.png')

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    prompt = "photo of an astronaut standing in a jungle, muted colors, detailed, 8k, realistic, helmet reflections"
    n_prompt = "blurry, ugly, low quality, jpeg compression, drawing, sketch, painting, cartoon, face"
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="pil",
        num_inference_steps=100, generator=random_generator, guidance_scale=6.5).images[0]
    image.save('spaceman100_1_other.png')

    random_generator = torch.Generator()
    random_generator.manual_seed(42)
    high_noise_frac = 0.8
    image = pipe(prompt=prompt, negative_prompt=n_prompt, output_type="latent" if use_refiner else "pil",
        denoising_end=high_noise_frac, num_inference_steps=100, generator=random_generator, guidance_scale=6.5).images[0]
    image = refiner(prompt=prompt, image=image[None, :], denoising_start=high_noise_frac, num_inference_steps=100,
        guidance_scale=6.5).images[0]
    image.save('spaceman100_2_other.png')
```

Here is the output, showing drastically different timesteps used for strength versus timestep selection. This is the first part where I use timesteps:

```
Loading pipeline components...: 100%|█████████████| 7/7 [00:03<00:00,  2.04it/s]
Loading pipeline components...: 100%|█████████████| 5/5 [00:01<00:00,  2.58it/s]
  0%|                                                   | 0/100 [00:00<?, ?it/s]timesteps tensor([999, 992, 985, 978, 971, 963, 956, 948, 940, 933, 924, 916, 908, 899,
        891, 882, 873, 864, 854, 845, 835, 825, 815, 805, 794, 783, 772, 761,
        749, 737, 725, 713, 700, 687, 674, 660, 646, 632, 618, 603, 587, 572,
        556, 540, 523, 506, 489, 472, 454, 436, 418, 400, 382, 364, 345, 327,
        309, 291, 273, 256, 239, 222, 205, 190, 174, 160, 146, 133, 120, 108,
         97,  87,  78,  69,  61,  53,  47,  41,  35,  31,  26,  23,  19,  16,
         14,  12,  10,   8,   7,   6,   4,   3,   2,   1,   0],
       device='cuda:0')
 95%|███████████████████████████████████████▉  | 95/100 [00:45<00:02,  2.09it/s]
  0%|                                                    | 0/63 [00:00<?, ?it/s]timesteps tensor([999, 992, 985, 978, 971, 963, 956, 948, 940, 933, 924, 916, 908, 899,
        891, 882, 873, 864, 854, 845, 835, 825, 815, 805, 794, 783, 772, 761,
        749, 737, 725, 713, 700, 687, 674, 660, 646, 632, 618, 603, 587, 572,
        556, 540, 523, 506, 489, 472, 454, 436, 418, 400, 382, 364, 345, 327,
        309, 291, 273, 256, 239, 222, 205], device='cuda:0')
100%|███████████████████████████████████████████| 63/63 [00:36<00:00,  1.70it/s]
  0%|                                                    | 0/32 [00:00<?, ?it/s]timesteps tensor([190, 174, 160, 146, 133, 120, 108,  97,  87,  78,  69,  61,  53,  47,
         41,  35,  31,  26,  23,  19,  16,  14,  12,  10,   8,   7,   6,   4,
          3,   2,   1,   0])
100%|███████████████████████████████████████████| 32/32 [00:22<00:00,  1.43it/s]
```

Here is the second part where I use high_noise_frac.

```
  0%|                                                   | 0/100 [00:00<?, ?it/s]timesteps tensor([999, 992, 985, 978, 971, 963, 956, 948, 940, 933, 924, 916, 908, 899,
        891, 882, 873, 864, 854, 845, 835, 825, 815, 805, 794, 783, 772, 761,
        749, 737, 725, 713, 700, 687, 674, 660, 646, 632, 618, 603, 587, 572,
        556, 540, 523, 506, 489, 472, 454, 436, 418, 400, 382, 364, 345, 327,
        309, 291, 273, 256, 239, 222, 205, 190, 174, 160, 146, 133, 120, 108,
         97,  87,  78,  69,  61,  53,  47,  41,  35,  31,  26,  23,  19,  16,
         14,  12,  10,   8,   7,   6,   4,   3,   2,   1,   0],
       device='cuda:0')
 95%|███████████████████████████████████████▉  | 95/100 [00:38<00:02,  2.48it/s]
  0%|                                                    | 0/80 [00:00<?, ?it/s]timesteps tensor([999, 992, 985, 978, 971, 963, 956, 948, 940, 933, 924, 916, 908, 899,
        891, 882, 873, 864, 854, 845, 835, 825, 815, 805, 794, 783, 772, 761,
        749, 737, 725, 713, 700, 687, 674, 660, 646, 632, 618, 603, 587, 572,
        556, 540, 523, 506, 489, 472, 454, 436, 418, 400, 382, 364, 345, 327,
        309, 291, 273, 256, 239, 222, 205, 190, 174, 160, 146, 133, 120, 108,
         97,  87,  78,  69,  61,  53,  47,  41,  35,  31], device='cuda:0')
100%|███████████████████████████████████████████| 80/80 [00:28<00:00,  2.77it/s]
  0%|                                                    | 0/20 [00:00<?, ?it/s]timesteps tensor([26, 23, 19, 16, 14, 12, 10,  8,  7,  6,  4,  3,  2,  1,  0],
       device='cuda:0')
 75%|████████████████████████████████▎          | 15/20 [00:06<00:02,  2.29it/s]
```

high_noise_frac is using totally different timesteps for the refiner unet! In this case, with `high_noise_frac` we use `[26, 23, 19, 16, 14, 12, 10,  8,  7,  6,  4,  3,  2,  1,  0]`, whereas with the timestep cutoff of 200 we use `190, 174, 160, 146, 133, 120, 108,  97,  87,  78,  69,  61,  53,  47, 41,  35,  31,  26,  23,  19,  16,  14,  12,  10,   8,   7,   6,   4, 3,   2,   1,   0`. In fact, you are using the refiner so little that you may as well not even switch to it!

What you are recommending instead (fraction) is a misunderstanding of both the training of Stable Diffusion XL's refiner and of how an ensemble of experts works.

AmericanPresidentJimmyCarter commented:
And yes, there is 100% a difference between these images! Here is your propose "0.8 strength cutoff":

![spaceman100_2_other](https://github.com/huggingface/diffusers/assets/110263573/3e8b47f5-ea26-4996-a171-243aa1f69e73)

And here is the proper image using the timestep 200 cutoff:

![spaceman100_2](https://github.com/huggingface/diffusers/assets/110263573/7092fc58-9e66-49c6-a5db-aeea6068f145)

As you can seen, when the refiner model is used _correctly_ it is easily able to fill in the American flag patch on the suit.

patrickvonplaten commented:
Thanks for the great investigation @AmericanPresidentJimmyCarter! I understand the issue now fully. The problem is that `use_karras_sigmas=True` and procentage-wise start and end denoising doesn't work because `use_karras_sigmas=True` distributes the timesteps not equally.

So just to summarize everything here.

1.) The refiner was trained only on the timesteps between 0 - 200.
2.) Therefore it all/most cases we should try to make sure that the refiner sees timesteps exactly in this range
3.) This works well if `use_karras_sigmas=False` but **not** if `use_karras_sigmas=True`. 
4.) Therefore to allow better support for `use_karras_sigmas=True` we should also allow int values to be passed

Would you agree here? 

Also just to verify this could you run the above code with `use_karras_sigms=False` and a procentual 0.8 cut-off?

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -788,9 +789,12 @@ def __call__(
             timesteps = timesteps[: num_warmup_steps + self.scheduler.order * num_inference_steps]
+        if denoising_end is not None and type(denoising_end) == int:
+            num_inference_steps = len(list(filter(lambda ts: ts >= denoising_end, timesteps)))
+            timesteps = timesteps[:num_inference_steps]
```
This works for me!

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -788,9 +789,12 @@ def __call__(
+        if denoising_end is not None and type(denoising_end) == float:
             num_inference_steps = int(round(denoising_end * num_inference_steps))
             timesteps = timesteps[: num_warmup_steps + self.scheduler.order * num_inference_steps]
+        if denoising_end is not None and type(denoising_end) == int:
```
```suggestion
        elif denoising_end is not None and type(denoising_end) == int:
```
easier to read

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:
```diff
@@ -523,11 +523,19 @@ def get_timesteps(self, num_inference_steps, strength, device, denoising_start=N
+        # that is, strength is determined by the denoising_start instead.
+        if type(denoising_start) == int:
+            timesteps = list(filter(lambda ts: ts < denoising_start, timesteps))
+            return torch.tensor(timesteps), len(timesteps)
```
ok for me

patrickvonplaten commented on docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:
```diff
@@ -180,17 +265,17 @@ Let's have a look at the image
-The ensemble-of-experts method works well on all available schedulers!
-
-</Tip>
+Note that for the Stable Diffusion XL base and refiner models, the best choice is to use the 
+pipeline described in example 1.) because it works with a multitude of schedulers. Using strength 
+to transition denoising between ensembles is best when using different Stable Diffusion models and
+different schedulers, for example Stable Diffusion XL base for the first expert and a finetuned SD
+1.x model for the refinement expert.
```
```suggestion
<Tip>

The ensemble-of-experts method works well on all available schedulers!

</Tip>
```

I would leave this for now. To me it works just fine - it's just different 20% in this case doesn't mean [1-200] but just the final 20% of timesteps which is exactly what we want.

patrickvonplaten commented on docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:
```diff
@@ -118,21 +118,106 @@ a couple community contributors which also helped shape the following `diffusers
+- [AmericanPresidentJimmyCarter](https://github.com/AmericanPresidentJimmyCarter)
 
-#### 1.) Ensemble of Expert Denoisers
+#### 1.) Ensemble of Expert Denoisers (Timesteps)
```
Can we leave this section as is since we don't use `use_karras_sigmas=True` here. It would be great if we could just convey that `denoising_start` and `denoising_end` can be specified with both `int` and `float` and state each other's advantages. I don't think we should make two full sections here.

patrickvonplaten commented:
@AmericanPresidentJimmyCarter,

Could we leave the docs compact. I think it's nice to allow the user to provide both `int` and `float` but I don't think we need a whole new section for this. Instead let's maybe leave everything in one section and explain to the user how they two are different

AmericanPresidentJimmyCarter commented:
OK, I have pushed a new commit with the changes you have requested.

I removed the old lion image pair (we probably don't need two if you want to keep it compact!), and replaced it with the newly generated pair using the `denoising_end`/`denoising_start` pair of value 200.

I have added the rationale of why you should use either 200 or 0.8 for the `denoising_end` transition point to the documentation, and also added multiple warnings specifying not to use this if you have set `use_karras_sigmas=True`.

patrickvonplaten reviewed (approved):
Very cool, thanks for iterating and bearing with me to be on the same page! 

@sayakpaul @yiyixuxu could you also give this a review?

sayakpaul commented on docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:
```diff
@@ -156,31 +164,52 @@ refiner = DiffusionPipeline.from_pretrained(
-constant. Also remember that the output of the base model should be in latent space:
+Stable Diffusion XL base is trained on timesteps 0-999 and Stable Diffusion refiner is finetuned
+from the base on timesteps 0-200, so we use the base model for the first 800 timesteps (high noise) 
+and the refiner for the last 200 timesteps (low noise). Hence, `high_noise_transition`
+is set to 200, so that all steps 200-999 are performed by the base model and steps 0-199 are
+performed by the refiner model.
```
We mention that the first 800 steps will be performed by the base model but here we're specifying 200-999, which are not the first 800 steps? 

What am I missing out on?

AmericanPresidentJimmyCarter replied:
Inference works backwards through the noising process, so timesteps 999 -> 0, iteratively. 999 = highest noise, 0 = lowest noise.

sayakpaul replied:
My bad. Sorry.

sayakpaul commented on docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:
```diff
@@ -156,31 +164,52 @@ refiner = DiffusionPipeline.from_pretrained(
+is set to 200, so that all steps 200-999 are performed by the base model and steps 0-199 are
+performed by the refiner model.
+
+Alternatively, a fraction of 0.8 (`high_noise_transition`) means that 80% of the 40 inference steps (32 steps) are
+run through the base model and the remaining 8 steps are run through the refiner when
+`use_karras_sigmas=False`, with the base model handling timesteps 201-999 while the refinement
+model handles timesteps 0-200. **Please note this timestep proportional relationship does not hold true when
+using the `use_karras_sigmas` option and you should instead use an integer value as described above.**
```
Hmm I see that there's a difference of 1 in the timesteps. In the "int" setting, the brackets are (0-199) and (200 - 999) but here it's (0, 200) and (201-999).

AmericanPresidentJimmyCarter replied:
Correct, see: https://github.com/huggingface/diffusers/pull/4115#issuecomment-1638433779

IMO: the float proportion is buggy, but since you already merged it I didn't want to make backwards incompatible changes.

sayakpaul commented on tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py:
```diff
@@ -223,7 +223,589 @@ def test_stable_diffusion_xl_offloads(self):
+            # Let's retrieve the number of timesteps we want to use
+            pipe_1.scheduler.set_timesteps(num_steps)
+            expected_steps = pipe_1.scheduler.timesteps.tolist()
+            print(scheduler_cls_orig)
+            print(expected_steps)
```
Can be safely deleted.

AmericanPresidentJimmyCarter replied:
Thanks, missed these.

sayakpaul commented on tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py:
```diff
@@ -223,7 +223,589 @@ def test_stable_diffusion_xl_offloads(self):
+
+            expected_steps_1 = list(filter(lambda ts: ts >= split, expected_tss))
+            expected_steps_2 = list(filter(lambda ts: ts < split, expected_tss))
+            print(expected_steps_1)
+            print(expected_steps_2)
```
Can be safely deleted.

AmericanPresidentJimmyCarter replied:
Thanks, missed these.

AmericanPresidentJimmyCarter commented on docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.mdx:
```diff
@@ -127,11 +128,18 @@ expert for the high-noise diffusion stage and the refiner serves as the expert f
+denoising stage (*i.e.* the refiner model) respectively. This should be set as the [`denoising_end`](https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/stable_diffusion_xl#diffusers.StableDiffusionXLPipeline.__call__.denoising_end) of the base model 
 and as the [`denoising_start`](https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/stable_diffusion_xl#diffusers.StableDiffusionXLImg2ImgPipeline.__call__.denoising_start) of the refiner model.
 
+For both `denoising_end` and `denoising_start`, either an integer or a float value may be passed. If an integer
+value is passed, the denoising end and start for both models will be defined in terms of discrete timesteps. If
+a float value is passed, the end and start of denoising will be defined by proportions of timesteps given
+the number of inference steps.
```
Could you elaborate on what sort of examples you would like to see, that are not included in the text below it? Do I need a longer explanation of timesteps?

sayakpaul replied:
Just numerical examples. Like:

"If you have a total of 40 inference steps, then specifying a value of 0.8 as `denoising_end` would mean ..."

But I think you already do this in the text below.

AmericanPresidentJimmyCarter commented:
> It'd be nice to add a test to the behavior of what happens when [this is](https://github.com/huggingface/diffusers/pull/4115/files#diff-87969947ac09db062b29b1f7285246c055f9219b34b89c871884e5feee2107e7R1204-R1206) violated.

Sure, I can look into this later tonight.

yiyixuxu reviewed (commented):
I think we should keep our API as it is (`denoising_start/denoising_end` will accept fraction only) and we should just handle it differently on our end when `use_karras_sigmas=True`

AmericanPresidentJimmyCarter commented:
> think we should keep our API as it is (`denoising_start/denoising_end` will accept fraction only - and that fraction should reflect how you divide training steps, not inference steps), and we should handle it differently on our end when `use_karras_sigmas=True`
> 
> actually, it doesn't even matter`use_karras_sigmas=True` or not, we can just always use `num_training_steps * denoising_end` to filter timesteps

s/training steps/discrete timesteps?

So you are in favor of using integer timesteps, but using a float to determine the integer timesteps, and fixing the existing off by one bug using the float * timesteps?

@sayakpaul @patrickvonplaten Please correspond internally and resolve this issue before I continue writing code.

AmericanPresidentJimmyCarter commented:
To me it also makes sense to preserve the float, since some diffusion models e.g. Paella are trained with continuous timesteps (just a float32 or float64 value), and so the range 0. ... 1. indicating the level of noise clearly applies to them.

Describing discrete integer cutoffs for models that are trained in discrete timesteps with float values also feels odd -- so I am in favor of the overload with `int`.

yiyixuxu commented:
@AmericanPresidentJimmyCarter 
> So you are in favor of using integer timesteps, but using a float to determine the integer timesteps, and fixing the existing off by one bug using the float * timesteps?

I'm in favor of letting the user determine the intervals based on a percentage of training steps - so yes, naturally, it is a float if we talk about the data type. 

More importantly, it should be a "fraction" of the __training steps__ (currently it's defined as a fraction of inference steps, which is problematic). And we use this fraction to determine the cutoff point for timesteps, for both integral and float timesteps.

AmericanPresidentJimmyCarter commented:
> I'm in favor of letting the user determine the intervals based on a percentage of training steps - so yes, naturally, it is a float if we talk about the data type.

This is fine, but as I have pointed out, with certain models this will lead to potentially having to input irrational numbers (as described above). Hence for models with discrete timesteps, an integer makes more sense.

AmericanPresidentJimmyCarter commented:
@yiyixuxu Actually, why don't I just fix the off-by-one bug, leave both `float` and `int`, and make the float proportion refer to a proportion of discrete timesteps?

Then users can select either.

patrickvonplaten commented:
Hey @AmericanPresidentJimmyCarter,

Sorry to only get back now. After some thinking it actually looks like @yiyixuxu hits the right argument here. If we just base `denoising_start` and `denoising_end` purely on the `num_training_steps` and only allow float values (as is) we should get the best of both worlds (easy to understand API), Also as you say it's much more future proof in case `num_training_steps` is not 0-1000 but 0-1 . Would it be ok for you to go this way? So:

- 1. **Not** allowing int values for `denoising_start` and `denoising_end` and
- 2. Refactor the mechanism we have now to base everything on `num_training_steps` (essentially using your filter mechanism, but just retrieving the 800 by `int(round(denoising_send * num_training_steps))`

AmericanPresidentJimmyCarter commented:
As I noted, it still doesn't make sense to not offer the integer cutoff for me, since you might decide to train an expert that doesn't fall into a neat value between 0 and 1 like `0.8`.

patrickvonplaten commented:
> As I noted, it still doesn't make sense to not offer the integer cutoff for me, since you might decide to train an expert that doesn't fall into a neat value between 0 and 1 like `0.8`.

Happy to revisit the decision when a new checkpoint is trained that doesn't have a clear cut. For now, I'd say the pro's of only allowing float values to be passed (much simpler & easier to understand) outweigh the scenario of a future unevenly trained model. 

Could we maybe for now take this road: https://github.com/huggingface/diffusers/pull/4115#issuecomment-1642471898 ? Also happy to finish this PR in that way

AmericanPresidentJimmyCarter commented:
OK, I will look at altering this so it uses a float to compute an integer.

AmericanPresidentJimmyCarter commented:
Rebased to main and pushed.

- @yiyixuxu @patrickvonplaten Only the float value is maintained, and the documentation has been updated accordingly.
- The off-by-one bug previously mentioned is fixed.
- Removed all references to the cutoffs setting the number of inference steps based on their proportions, since this isn't true for kerras sigmas and after the off-by-one bug was fixed.
- @sayakpaul The ValueError never tested before in img2img and inpaint pipelines is now tested against.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -599,12 +599,11 @@ def __call__(
-                denoising steps. As a result, the returned sample will still retain a substantial amount of noise. The
-                denoising_end parameter should ideally be utilized when this pipeline forms a part of a "Mixture of
-                Denoisers" multi-pipeline setup, as elaborated in [**Refining the Image
+                As a float: When specified, determines the fraction (between 0.0 and 1.0) of the total denoising
```
```suggestion
                When specified, determines the fraction (between 0.0 and 1.0) of the total denoising
```

patrickvonplaten commented:
Thanks for iterating on this one @AmericanPresidentJimmyCarter
